decision_service/rlclientlib/restapi_data_transport.cc

```diff
@@ -1,10 +1,138 @@
+   * x-ms-server-encrypted = true
+   * x-ms-version = 2017-04-17
+   */
+
+  int restapi_data_tranport::get_data_info(::utility::datetime& last_modified, ::utility::size64_t& sz, api_status* status) {
+
+    // Build request URI and start the request.
+    pplx::task<int> requestTask = _httpcli.request(methods::HEAD)
+      // Handle response headers arriving.
+      .then([&](http_response response) {
+      if ( response.status_code() != 200 )
+        RETURN_ERROR(status, error_code::http_bad_status_code, error_code::http_bad_status_code_s);
+
+      auto iter = response.headers().find(U("Last-Modified"));
+      if(iter == response.headers().end())
+        RETURN_ERROR(status, error_code::last_modified_not_found, error_code::last_modified_not_found_s);
+
+      last_modified = ::utility::datetime::from_string(iter->second);
+      if( last_modified.to_interval() == 0)
+        RETURN_ERROR(status, error_code::last_modified_invalid, error_code::last_modified_invalid_s);
+
+      sz = response.headers().content_length();
+      if( sz <= 0 )
+        RETURN_ERROR(status, error_code::bad_content_length, error_code::bad_content_length_s);
+
+      return error_code::success;
+    });
+
+    // Wait for all the outstanding I/O to complete and handle any exceptions
+    try {
+      requestTask.wait();
+    }
+    catch ( const std::exception &e ) {
+      RETURN_ERROR(status,error_code::exception_during_http_req,e.what());
+    }
+
+    return error_code::success;
+  }
+
+  int restapi_data_tranport::get_data(model_data& ret, api_status* status) {
+
+    ::utility::datetime curr_last_modified;
+    ::utility::size64_t curr_datasz;
+    auto err = get_data_info(curr_last_modified, curr_datasz, status);
+    TRY_OR_RETURN(err);
+
+    if ( curr_last_modified == _last_modified && curr_datasz == _datasz )
+      return error_code::success;
+
+    // Build request URI and start the request.
+    pplx::task<int> requestTask = _httpcli.request(methods::GET)
+      // Handle response headers arriving.
+      .then([&](pplx::task<http_response> respTask) {
+      auto response = respTask.get();
+      if ( response.status_code() != 200 )
+        RETURN_ERROR(status, error_code::http_bad_status_code, error_code::http_bad_status_code_s);
+
+      auto iter = response.headers().find(U("Last-Modified"));
+      if ( iter == response.headers().end() )
+        RETURN_ERROR(status, error_code::last_modified_not_found, error_code::last_modified_not_found_s);
+
+      curr_last_modified = ::utility::datetime::from_string(iter->second);
+      if ( curr_last_modified.to_interval() == 0 )
+        RETURN_ERROR(status, error_code::last_modified_invalid, error_code::last_modified_invalid_s);
+
+      curr_datasz = response.headers().content_length();
+      if ( curr_datasz <= 0 )
+        RETURN_ERROR(status, error_code::bad_content_length, error_code::bad_content_length_s);
+
+      const auto buff = new char[curr_datasz];
+      ret.data = buff;
```

Comment:
what happens to this memory in case response.body().read_to_end(rb).get throws?

Reply:
Good point.  Fixed

---
In reply to: [193495029](https://github.com/JohnLangford/vowpal_wabbit/pull/1495#discussion_r193495029) [](ancestors = 193495029)